File: android/app/src/main/java/com/dam/starwarsapp/domain/repository/VideoProvider.kt
package com.dam.starwarsapp.domain.repository

import com.dam.starwarsapp.domain.model.VideoCandidate
import com.dam.starwarsapp.domain.model.PlaybackTarget

/**
 * Provider abstraction for searching and resolving videos.
 */
interface VideoProvider {
    val id: String
    suspend fun searchFirst(title: String): VideoCandidate?
    suspend fun resolvePlayback(candidate: VideoCandidate): PlaybackTarget?
}

File: iOS/Star-Wars-Demo-APP/Star-Wars-Demo-APP/Domain/Service/VideoProvider.swift
import Foundation

protocol VideoProvider {
    func searchFirst(title: String) async throws -> VideoCandidate?
    func resolvePlayback(for candidate: VideoCandidate) async throws -> PlaybackTarget?
}

protocol VideoResolver {
    /// Resolve a title into a candidate and playback target. Returns nil if no provider could resolve.
    func resolveVideo(title: String) async throws -> (candidate: VideoCandidate, target: PlaybackTarget)?
}
